runtime/shibuya/src/xcm_config.rs

```diff
@@ -186,6 +194,120 @@ impl<T: Get<MultiLocation>> FilterAssetLocation for AssetsFrom<T> {
 }
 */
 
+pub struct FixedRateOfForeignAsset<T: ExecutionPaymentRate, R: TakeRevenue> {
+    /// Total used weight
+    weight: Weight,
+    /// Total consumed assets
+    consumed: u128,
+    /// Asset Id (as MultiLocation) and units per second for payment
+    asset_location_and_units_per_second: Option<(MultiLocation, u128)>,
+    _pd: PhantomData<(T, R)>,
+}
+
+impl<T: ExecutionPaymentRate, R: TakeRevenue> WeightTrader for FixedRateOfForeignAsset<T, R> {
+    fn new() -> Self {
+        Self {
+            weight: 0,
+            consumed: 0,
+            asset_location_and_units_per_second: None,
+            _pd: PhantomData,
+        }
+    }
+
+    fn buy_weight(
+        &mut self,
+        weight: Weight,
+        payment: xcm_executor::Assets,
+    ) -> Result<xcm_executor::Assets, XcmError> {
+        log::trace!(
+            target: "xcm::weight",
+            "FixedRateOfForeignAsset::buy_weight weight: {:?}, payment: {:?}",
+            weight, payment,
+        );
+
+        // Atm in pallet, we only support one asset so this should work
+        let payment_asset = payment
+            .fungible_assets_iter()
+            .next()
+            .ok_or(XcmError::TooExpensive)?;
+
+        match payment_asset {
+            MultiAsset {
+                id: xcm::latest::AssetId::Concrete(asset_location),
+                // sanity check but can be omited? TODO
+                fun: Fungibility::Fungible(_),
+            } => {
+                // TODO: have to account for the DB read somewhere in weight. Or already embeded into assumed xcm instruction weight?
+                if let Some(units_per_second) = T::get_units_per_second(asset_location.clone()) {
+                    let amount = units_per_second * (weight as u128) / (WEIGHT_PER_SECOND as u128);
+                    if amount == 0 {
+                        return Ok(payment);
+                    }
+
+                    let unused = payment
+                        .checked_sub((asset_location.clone(), amount).into())
+                        .map_err(|_| XcmError::TooExpensive)?;
+
+                    self.weight = self.weight.saturating_add(weight);
+
+                    // If there are multiple calls to `BuyExecution` but with different assets, we need to be able to handle that.
+                    // Current primitive implementation will just keep total track of consumed asset for the FIRST consumed asset.
+                    // Others will just be ignored when refund is concerned.
+                    // TODO: improve this, now or via backlog item?
```

Comment:
I will create a backlog item to improve this.
Will most likely be required in the future.